### PR TITLE
implement assert_tzawareness_compat for DatetimeIndex

### LINKED ISSUE #18162: DatetimeIndex comparison tzaware vs naive should raise
```
>>> dr = pd.date_range('2016-01-01', periods=6)
>>> dz = dr.tz_localize('US/Pacific')
>>> dr < dz
array([ True,  True,  True,  True,  True,  True], dtype=bool)
>>> dr[0] < dz[0]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/_libs/tslib.pyx", line 1169, in pandas._libs.tslib._Timestamp.__richcmp__
  File "pandas/_libs/tslib.pyx", line 1230, in pandas._libs.tslib._Timestamp._assert_tzawareness_compat
TypeError: Cannot compare tz-naive and tz-aware timestamps
```

The vectorized comparison should raise too right?

### PR DESCRIPTION
**Pull request by jbrockmendel**
closes #18162 

ATM comparing tzaware `DatetimeIndex` with tznaive DTI fails to raise.  This PR implements `_assert_tzawareness_compa`t (which currently exists in `_Timestamp`) in `DatetimeIndex` to fix that.

That in turn causes breakage in `Series.replace`.  There's a small edit in `core.internals` to fix that, but I'm not sure that's the best way to make that work.

There is still one remaining test error in `tests.series.test_indexing.TestSeriesIndexing.test_getitem_setitem_datetimeindex`, which may either represent a bug or a case where tznaive/tzaware rules are relaxed.

```
N = 50
rng = date_range('1/1/1990', periods=N, freq='H', tz='US/Eastern')
ts = Series(np.random.randn(N), index=rng)

lb = "1990-01-01 04:00:00"
rb = "1990-01-01 07:00:00"

>>> result = ts[(ts.index >= lb) & (ts.index <= rb)]
TypeError: Cannot compare tz-naive and tz-aware datetime-like objects
```

Later in the same tests it tries to compare `ts.index` to naive `datetime` objects.  Is this a special case where we are intentionally less strict?

### CODE REVIEW

File: pandas/core/indexes/datetimes.py [hunk 1]
```diff
@@ -649,6 +653,20 @@ def _simple_new(cls, values, name=None, freq=None, tz=None,
         result._reset_identity()
         return result
 
+    def _assert_tzawareness_compat(self, other):
+        # adapted from _Timestamp._assert_tzawareness_compat
+        other_tz = getattr(other, 'tzinfo', None)
```
Comment (jreback) [inline pandas/core/indexes/datetimes.py]: this is odd to use ``tzinfo``, these are already wrapped scalars, or an index type, so ``.tz`` is appropriate

File: pandas/core/internals.py [hunk 2]
```diff
@@ -3486,10 +3486,20 @@ def replace_list(self, src_list, dest_list, inplace=False, regex=False,
         # figure out our mask a-priori to avoid repeated replacements
         values = self.as_matrix()
 
-        def comp(s):
-            if isna(s):
-                return isna(values)
-            return _maybe_compare(values, getattr(s, 'asm8', s), operator.eq)
+        if is_datetime64tz_dtype(self):
```
Comment (jreback) [inline pandas/core/internals.py]: this is not the right place to fix if this is even an issue

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/datetimes.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimes.py]: That was my first thought too but `other` could be a raw `datetime`.

File: pandas/core/internals.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/internals.py]: I tend to agree, could use advice on where _is_ the right place.  Without this change there is a test failure in tests.indexing.test_coercion.TestReplaceSeriesCoercion.test_replace_series_datetime64tz with traceback

```
    def test_replace_series_datetime64tz(self):
        from_key = 'datetime64[ns, US/Eastern]'
        for to_key in self.rep:
>           self._assert_replace_conversion(from_key, to_key, how='dict')

pandas/tests/indexing/test_coercion.py:1317: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pandas/tests/indexing/test_coercion.py:1239: in _assert_replace_conversion
    result = obj.replace(replacer)
pandas/core/generic.py:4514: in replace
    limit=limit, regex=regex)
pandas/core/generic.py:4563: in replace
    regex=regex)
pandas/core/internals.py:3504: in replace_list
    masks = [comp(s) for i, s in enumerate(src_list)]
pandas/core/internals.py:3502: in comp
    operator.eq)
pandas/core/internals.py:4954: in _maybe_compare
    result = op(a, b)
pandas/core/indexes/datetimes.py:122: in wrapper
    self._assert_tzawareness_compat(other)
```

where the relevant `obj.replace(replacer)` call has

```
index = pd.Index([3, 4], name='xxx')
data = [Timestamp('2011-01-01 00:00:00-0500', tz='US/Eastern'), Timestamp('2011-01-03 00:00:00-0500', tz='US/Eastern')]
obj = pd.Series(data, index=index, name='yyy')
replacer = {Timestamp('2011-01-01 00:00:00-0500', tz='US/Eastern'): 1.1, Timestamp('2011-01-03 00:00:00-0500', tz='US/Eastern'): 2.2}
```

The problem is that in internals `replace_list` is converting to m8 which drops tz.

**Comment by jbrockmendel** [comment]
#17920 makes some desig decisions about tzawareness strictness.   Conditional on those decisions, comparisons should probably use the same conventions.

**Comment by jreback** [comment]
#17920 is orthogonal. That is purely string based. When you use Timestamps, they *must* be the same timezone (or None) or should raise

**Comment by jbrockmendel** [comment]
> #17920 is orthogonal. That is purely string based. When you use Timestamps, they must be the same timezone (or None) or should raise

See the OP.  The unresolved error here is in comparing a tzaware DatetimeIndex against naive-like strings.


```
rng = pd.date_range('1/1/1990', periods=50, freq='H', tz='US/Eastern')
ts = pd.Series(np.random.randn(50), index=rng)

lb = "1990-01-01 04:00:00"
rb = "1990-01-01 07:00:00"
```

Under the status quo, the following are equivalent:

```
ts[lb:rb]
ts[(ts.index >= lb) & (ts.index <= rb)]
```

#17920 affects the behavior of `ts[lb:rb]` (though the output is unchanged in this case) while this affects the behavior of `ts.index >= lb`.  If we fix the comparison issue and implement #17920, we break the equivalence of slicing methods.  Which may be acceptable; I just want to be explicit about it.

**Comment by 1kastner** [comment]
@jbrockmendel So how do you think this issue could be solved? Honestly I used some parts of pandas for my project but I do not have deep insight into the internals and wherelse the changes I suggested could be made. Or is it rather that the idea behind my changes need to be implemented at more places for a coherent behavior? If so, which places do you have in mind?

**Comment by jbrockmendel** [comment]
Increasingly I think we should be strict about this, i.e. be Technically Correct.  Anything else is going to require keeping track of what rules are loosened in which special cases, and will inevitably lead to headaches down the road.

That said, an idea for a workaround for slicing tzaware indexes: interpret a trailing "TZ" at the end of a string as "interpret this naive datetime-like string with this DatetimeIndex's timezone".

Is #17920 motivated by a pressing use case where you've got a tzaware DatetimeIndex?  I expect the large majority of indexing use cases are tznaive.

**Comment by 1kastner** [comment]
It is motivated by a use case with data from several different data sources. Some are UTC, some are German Winter Time (without daylight saving time, definitely not a standard time), some are CET etc. I believe that mixing different data sources often has timezone issues as a consequence. But a prove I can not serve.

Adding a `TZ` is not an option in ISO 8601 as far as I have read it. And I would try to comply to that standard for the labels. Your attempt sounds quite experimential to me.

**Comment by jbrockmendel** [comment]
Let's move the discussion over to #18435.  Closing this PR until the design issue is resolved.

**Comment by jbrockmendel** [comment]
Reopening.  If we're going to introduce inconsistencies, might as well go full-speed ahead and get them sorted out sooner rather than later.  Will push update shortly.

File: pandas/tests/indexes/datetimes/test_datetime.py [hunk 3]
```diff
@@ -248,6 +250,36 @@ def test_append_join_nondatetimeindex(self):
...
+        ops = [operator.eq, operator.ne,
+               operator.gt, operator.ge,
+               operator.lt, operator.le]
+
+        for left, right in [(dr, dz), (dz, dr)]:
+            for op in ops:
+                with pytest.raises(TypeError):
+                    op(left, right)
+
+        # Check that there isn't a problem aware-aware and naive-naive do not
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_datetime.py]: these tests should be in series (the latter ones). alternatively a better place might be tin test_base for these, where we already handle series & index tests.

File: pandas/core/indexes/datetimes.py (hunk 1 above)
Comment (jreback) [reply pandas/core/indexes/datetimes.py]: use .tz, you can simply wrap ``other = pd.Timestamp(other)``

File: pandas/core/internals.py (hunk 2 above)
Comment (jreback) [reply pandas/core/internals.py]: this is not the place for this, might need to fixe ``.replace`` in the block itself.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
> this is not the place for this, might need to fixe .replace in the block itself.

The pertinent call doesn't go through Block.replace.  NDFrame.replace calls BlockManager.replace_list.  Maybe edit NDFrame.replace to call Block.replace and add the appropriate check to DatetimeBlock.replace?  Whatever we can do to limit the scope...

**Comment by jbrockmendel** [comment]
If we don't want to special-case the `comp` function, then we could dispatch to `b.replace` in like in the `b.dtype == np.object_` case, then override `replace` for `DatetimeTZBlock`.  Is that what you have in mind?

**Comment by jreback** [comment]
so ``replace_list`` is a giant hack ATM and needs to be fixed. maybe this is the impetus. However, don't really to add your hack on top. What exactly is failing w/o you touching replace? let's isolate and x-fail those tests for now.

**Comment by jbrockmendel** [comment]
> so replace_list is a giant hack ATM and needs to be fixed. maybe this is the impetus. However, don't really want to add your hack on top. What exactly is failing w/o you touching replace? let's isolate and x-fail those tests for now.

I like this plan, I'm [excited to be a part of it](https://www.youtube.com/watch?v=PVhTDNlbsSc).

In tests.indexing.test_coercion.TestReplaceSeriesCoercion.test_replace_series_datetime64tz, we have:

```
index = pd.Index([3, 4], name='xxx')
data = [Timestamp('2011-01-01 00:00:00-0500', tz='US/Eastern'), Timestamp('2011-01-03 00:00:00-0500', tz='US/Eastern')]
obj = pd.Series(data, index=index, name='yyy')
replacer = {Timestamp('2011-01-01 00:00:00-0500', tz='US/Eastern'): 1.1, Timestamp('2011-01-03 00:00:00-0500', tz='US/Eastern'): 2.2}

>>> obj.replace(replacer)  # <-- raises
```

When it gets to `replace_list`, it calls `asm8` on those `Timestamps`, which drops their tz information.  Then it compares against `values`, which at this point is a tzaware `DatetimeIndex`, so now raises a `TypeError` for lack of tzawareness-compat.

I'll have to re-run the whole suite to be sure, but I think that xfailing this one test will get to green without piling on hacks in internals.  I'll circle back to try to de-hack that after wrapping up the current set of PRs.

**Comment by jreback** [comment]
> I like this plan, I'm excited to be a part of it.

not sure if you remember [what happens next](https://www.youtube.com/watch?v=d-sALU_hveA)

**Comment by jbrockmendel** [comment]
Fortunately we all know what to do when asked if we’re gods.

File: pandas/tests/indexes/datetimes/test_datetime.py [hunk 4]
```diff
@@ -248,6 +250,49 @@ def test_append_join_nondatetimeindex(self):
         # it works
         rng.join(idx, how='outer')
 
+    def test_comparison_tzawareness_compat(self):
+        # GH#18162
+        dr = pd.date_range('2016-01-01', periods=6)
+        dz = dr.tz_localize('US/Pacific')
+
+        ops = [operator.eq, operator.ne,
+               operator.gt, operator.ge,
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_datetime.py]: parametrize this you can make multiple tests if thats easier

File: pandas/tests/indexes/test_base.py [hunk 5]
```diff
@@ -2240,6 +2240,28 @@ def test_intersect_str_dates(self):
...
+
+        ops = [operator.eq, operator.ne,
+               operator.gt, operator.ge,
+               operator.lt, operator.le]
+
+        # Check that there isn't a problem aware-aware and naive-naive do not
+        # raise
+        naive_series = Series(dr)
+        aware_series = Series(dz)
+        for op in ops:
```
Comment (jreback) [inline pandas/tests/indexes/test_base.py]: same

File: pandas/tests/indexing/test_coercion.py [hunk 6]
```diff
@@ -1311,6 +1311,7 @@ def test_replace_series_datetime64(self):
         for to_key in self.rep:
             self._assert_replace_conversion(from_key, to_key, how='series')
 
+    @pytest.mark.xfail
```
Comment (jreback) [inline pandas/tests/indexing/test_coercion.py]: need a reason, reference a specific issue

File: pandas/tests/series/test_indexing.py [hunk 7]
```diff
@@ -474,6 +481,13 @@ def test_getitem_setitem_datetimeindex(self):
 
         lb = datetime(1990, 1, 1, 4)
         rb = datetime(1990, 1, 1, 7)
+        with pytest.raises(TypeError):
+            # tznaive vs tzaware comparison is invalid
+            # see GH#18376, GH#18162
+            ts[(ts.index >= lb) & (ts.index <= rb)]
+
+        lb = rng.tzinfo.localize(datetime(1990, 1, 1, 4))
```
Comment (jreback) [inline pandas/tests/series/test_indexing.py]: don't use use tz specific routines, rather use ``tz_localize``

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jorisvandenbossche** [comment]
No longer interpreting a string in the timezone of the index it is being compared to (but raising an error), seem a big compatibility break to me.

**Comment by jbrockmendel** [comment]
@jorisvandenbossche you realize this makes you the staypuft of this conversation, right?

**Comment by jbrockmendel** [comment]
Does this need discussion on eg the mailing list?

**Comment by jbrockmendel** [comment]
Thoughts on this? If we knock this down I can look at replace_list.

File: pandas/core/internals.py [hunk 8]
```diff
@@ -3491,8 +3491,12 @@ def comp(s):
                 return isna(values)
             return _maybe_compare(values, getattr(s, 'asm8', s), operator.eq)
 
-        masks = [comp(s) for i, s in enumerate(src_list)]
-
+        try:
+            masks = [comp(s) for i, s in enumerate(src_list)]
+        except:
+            print(src_list)
```
Comment (jreback) [inline pandas/core/internals.py]: huh?

File: pandas/core/indexes/datetimes.py [hunk 9]
```diff
@@ -105,8 +105,11 @@ def _dt_index_cmp(opname, cls, nat_result=False):
 
     def wrapper(self, other):
         func = getattr(super(DatetimeIndex, self), opname)
-        if (isinstance(other, datetime) or
-                isinstance(other, compat.string_types)):
+        if isinstance(other, (datetime, compat.string_types)):
+            if isinstance(other, compat.string_types):
```
Comment (jreback) [inline pandas/core/indexes/datetimes.py]: would put this block outside the top-level if, then you know that you have datetimes

File: pandas/core/indexes/datetimes.py [hunk 10]
```diff
@@ -652,6 +656,20 @@ def _simple_new(cls, values, name=None, freq=None, tz=None,
         result._reset_identity()
         return result
 
+    def _assert_tzawareness_compat(self, other):
+        # adapted from _Timestamp._assert_tzawareness_compat
+        other_tz = getattr(other, 'tzinfo', None)
+        if is_datetime64tz_dtype(other):
+            # Get tzinfo from Series dtype
+            other_tz = other.dtype.tz
```
Comment (jreback) [inline pandas/core/indexes/datetimes.py]: I think would make sense to make a single routine in tslib/timezones.pyx which does the tzawareness check (make it a function), then call it both from Timestamp and here.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/internals.py (hunk 8 above)
Comment (jbrockmendel) [reply pandas/core/internals.py]: Leftover from debugging, will remove.  Good catch.

File: pandas/core/indexes/datetimes.py (hunk 10 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimes.py]: I see the appeal in this but have some reservations.  Do you mind saving this for the next round so I can give it some more thought?

**Comment by jbrockmendel** [comment]
Just pushed with edits per suggestions along with a whatsnew note.

File: doc/source/whatsnew/v0.22.0.txt [hunk 11]
```diff
@@ -320,5 +320,7 @@ Categorical
 Other
 ^^^^^
 
+- Improved error message when attempting to use a Python keyword as an identifier in a numexpr query (:issue:`18221`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: merging issues

File: pandas/tests/indexing/test_coercion.py [hunk 12]
```diff
@@ -1311,6 +1311,8 @@ def test_replace_series_datetime64(self):
         for to_key in self.rep:
             self._assert_replace_conversion(from_key, to_key, how='series')
 
+    # GH #18376, tzawareness-compat bug in BlockManager.replace_list
+    @pytest.mark.xfail
```
Comment (jreback) [inline pandas/tests/indexing/test_coercion.py]: instead of a comment, use the reason with the comment text

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
ok I pushed a merge of master to this. ping on green.

**Review by jreback — APPROVED** [review]

**Comment by jbrockmendel** [comment]
Travis timeout, appveyor `___ TestDataFrameGroupByPlots.test_series_groupby_plotting_nominally_works ____`

**Comment by jorisvandenbossche** [comment]
Sorry for not having further participated in the discussion, but repeating from my comment above: no longer interpreting a string in the timezone of the index it is being compared to (but raising an error), seems a big compatibility break to me. 

And I think we certainly should not regard it is as just a bug fix (eg in the whatsnew notes).

**Comment by jbrockmendel** [comment]
Ping

**Comment by jreback** [comment]
did u see joris comment

**Comment by jorisvandenbossche** [comment]
It is a bit easy to say for me as I am not doing the leg-work here, but I think it would be good if somebody made kind of an overview of the different related issues (this one, https://github.com/pandas-dev/pandas/issues/18435, https://github.com/pandas-dev/pandas/pull/17920, possibly others) related to timestamp comparisons: what is the current behaviour for the different cases (comparisons vs slicing, strings vs timestamp objects, scalars vs datetimeindex/series), where are the inconsistencies, what are the different options to solve this, ...

As I have the feeling that https://github.com/pandas-dev/pandas/pull/17920 is going in a different direction as this one (but difficult to see the full picture), and it would not be good to introduce more inconsistencies

Eg I think it would be OK to behave differently for comparing with strings vs timestamps: for strings interpret it in timezone it is being compared to, for timestamps raise if comparing non-aware with aware. But that is AFAIK not what this PR is doing.

**Comment by jbrockmendel** [comment]
> good if somebody made kind of an overview of the different related issues

That was the idea behind #18435.  Feel free to edit if that if the OP doesn't provide enough overview for your tastes.

> And I think we certainly should not regard it is as just a bug fix (eg in the whatsnew notes).

How to treat it in the whatsnew notes is above my pay grade.  But over the course of this PR's history I've become increasingly convinced that the current comparison behavior is Just Plain Wrong and should be treated like a bug.

The three options on hand are 1) this PR which makes `DatetimeIndex` comparisons behave like all other datetime-like comparisons, 2) edit the spec to explicitly make string comparisons a special case, 3) do nothing.  I'm going to assume away 3 and argue against 2.

AFAICT the main objection to 1) is that in conjunction with #17920 it breaks the equivalence between `ser.loc[lower:upper]` and `ser[(ser.index >= lower) & (ser.index <= upper)]`.  But consider what other equivalences are broken by 2:

- We can no longer count on `DatetimeIndex` comparisons to be transitive.  `a >= b` and `b >= c` no longer implies `a >= c`.
- Comparison-broadcasting becomes inconsistent.  `(index >= bound)[n]` no longer equals `index[n] >= bound` (unless we then go mess with `Timestamp`...)
- Comparison-boxing becomes inconsistent.  `index >= bound` no longer necessarily equals `index >= Timestamp(bound)`.
- Box-conversion becomes inconsistent. `index >= bound` no longer necessarily equals `index.astype(object) >= bound`

... and most of all, I am not remotely confident that this list is complete.  How many places across the code-base do comparisons with `DatetimeIndex` objects in one place or another?  I have no idea.

A tz-aware `DatetimeIndex` is not easy to get by accident.  If a user has one, they have made a decision that timezones matter.

-------
Any of the available options introduces an inconsistency _somewhere_.  AFAICT Option1 breaks a convenience equivalency, will do so loudly, and as a result will not snowball into other inconsistencies.

Special-casing string comparisons generates a whole mess of other potential (often silent) problems that can be avoided by enforcing behavior that is already canonical.

**Comment by jorisvandenbossche** [comment]
> That was the idea behind #18435. Feel free to edit if that if the OP doesn't provide enough overview for your tastes.

Ah, yes, forgot there was already quite some discussion there. Can you put your last comment there as well? Then I will answer there

**Comment by jbrockmendel** [comment]
Just pushed a commit that lets tzawareness compat slide for strings, enforces it for everything else.

**Comment by jbrockmendel** [comment]
@jorisvandenbossche with the most recent change this allows strings through without changing current behavior, only checks tzawareness-compat for datetime and vectors.  i.e. hopefully this fixes the part that we all agree is a bug without taking a stand on the rest.

File: pandas/tests/indexing/test_coercion.py [hunk 13]
```diff
@@ -821,6 +821,9 @@ def test_replace_series(self, how, to_key, from_key):
         if (from_key.startswith('datetime') and to_key.startswith('datetime')):
             pytest.xfail("different tz, currently mask_missing "
                          "raises SystemError")
+        elif from_key in ['datetime64[ns, US/Eastern]', 'datetime64[ns, UTC]']:
+            pytest.xfail(reason='GH #18376, tzawareness-compat bug '
+                                'in BlockManager.replace_list')
```
Comment (jorisvandenbossche) [inline pandas/tests/indexing/test_coercion.py]: might be discussed before, but since you need to add a xfail here, is this introducing a regression?

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks!

Should this also be tested for series/series and series/index comparisons?

File: pandas/tests/indexing/test_coercion.py (hunk 13 above)
Comment (jbrockmendel) [reply pandas/tests/indexing/test_coercion.py]: I guess?  See [above](https://github.com/pandas-dev/pandas/pull/18376#issuecomment-346969384):

> so replace_list is a giant hack ATM and needs to be fixed. maybe this is the impetus. However, don't really want to add your hack on top. What exactly is failing w/o you touching replace? let's isolate and x-fail those tests for now.

**Comment by jbrockmendel** [comment]
> Should this also be tested for series/series and series/index comparisons?

Separately.  I've just started looking at Series vs Index arithmetic/comparisons and that's going to be a long process (see #18824).  This PR has been a tough slog; I'd really like to get it over with.

**Comment by jreback** [comment]
this looks fine now that have eliminated the controversial string coercing.

@jorisvandenbossche

File: pandas/tests/indexing/test_coercion.py (hunk 13 above)
Comment (jorisvandenbossche) [reply pandas/tests/indexing/test_coercion.py]: Can you give an explicit code example that works now and will fail after this PR?

**Comment by jreback** [comment]
rebase

File: pandas/tests/indexing/test_coercion.py (hunk 13 above)
Comment (jbrockmendel) [reply pandas/tests/indexing/test_coercion.py]: @jorisvandenbossche good catch.  Two things here:

1) Does the fact that this is not immediately obvious by inspection suggest that test parametrization may have been taken a step too far?

2) In putting together an answer to this question I found that any non-datetime comparison raises, whereas I expect we want `DatetimeIndex(...) == "baz"` to just be Falsey (following the behavior of `Timestamp.__richcmp__` and `Timedelta.__richcmp__`).  So I need to fix this, will answer the original question after that.

Comment (jbrockmendel) [reply pandas/tests/indexing/test_coercion.py]: @jorisvandenbossche An example of a case that xfails after this PR:

```
index = pd.Index([3, 4], dtype='int64', name=u'xxx')
obj = pd.Series([pd.Timestamp('2011-01-01', tz='US/Eastern'),
                 pd.Timestamp('2011-01-03', tz='US/Eastern')],
                index=index, name='yyy')
replacer = pd.Series([pd.Timedelta(days=1), pd.Timedelta(days=2)],
                     index=obj)

>>> result = obj.replace(replacer)
[...]
TypeError: Cannot compare tz-naive and tz-aware datetime-like objects
```

File: doc/source/whatsnew/v0.23.0.txt [hunk 14]
```diff
@@ -395,4 +395,5 @@ Categorical
 Other
 ^^^^^
 
+- Fixed bug where comparing :class:`DatetimeIndex` failed to raise ``TypeError`` when attempting to compare timezone-aware and timezone-naive datetimelike objects (:issue:`18162`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: move to conversaion

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
whatsnew change, otherwise lgtm.

**Comment by jbrockmendel** [comment]
Ping

**Comment by jreback** [comment]
thanks!

**Pull request merged**